### DISCUSSION TITLE
XOXO export

### DISCUSSION THREAD

**Discussion by marrus-sh** [vincode-io/Zavala#208] [Ideas]
I’d propose adding an export to [XOXO (eXtensible Open XHTML Outlines)](http://microformats.org/wiki/xoxo), or some superset of it.

## Motivation

Right now both of the Zavala export formats require processing Markdown. I’d like an export format I can process with non‐Markdown‐aware XML tools, or publish directly to the Web.

## Proposed Export Format

Something like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!-- XOXO generated by Zavala -->
<html
  xmlns="http://www.w3.org/1999/xhtml"
  xmlns:opml="http://opml.org/spec2"
>
<head>
  <title>My Zavala Note</title>
  <meta name="generator" content="Zavala (https://zavala.vincode.io)"/>
  <!-- OPML metadata can be conveyed using the OPML namespace -->
  <opml:dateCreated>Wed, 10 Jan 2024 05:09:14 GMT</opml:dateCreated>
  <opml:dateModified>Wed, 10 Jan 2024 06:58:16 GMT</opml:dateModified>
  <opml:ownerName>marrus-sh</opml:ownerName>
  <opml:ownerEmail> marrus-sh@users.noreply.github.com</opml:ownerEmail>
  <opml:ownerID>https://github.com/marrus-sh</opml:ownerID>
</head>
<body>
  <ul class="xoxo">
    <li>My topic; <em>Markdown</em> is converted to <strong>HTML</strong>
      <ul>
        <li>My subtopic, with a note
          <dl>
            <dt>description</dt>
            <dd>In XOXO, properties of outline items are conveyed with a &lt;dl&gt; element.<br/>
"description" is a reasonable property for conveying notes.</dd>
          </dl>
          <ul>
            <li>Child items come after the metadata properties for an item</li>
          </ul>
        </li>
      </ul>
    </li>
    <li compact="compact">A compact topic
      <ul>
        <li>Hidden subtopic, because the parent is compact</li>
      </ul>
    </li>
  </ul>
</body>
</html>
```

**Comment by vincode-io** [comment]
XOXO is a 20 year old format that never caught on and isn't supported by any other outliner that I know of. Let me know if there is any support out there for it. I'm sorry, but I'd be much more likely to support something like Microsoft's docx format.

You can work around the Markdown issue you mention using Shortcuts. Just convert the Markdown output from Zavala to Rich Text and then convert that Rich Text to HTML. There are two built in Shortcut actions that do that. I use it with a custom script to publish blog posts that create as Zavala outlines to Jekyll.

I added some pretty extensive Shortcuts support to Zavala so that I could keep the main app as simple as possible. That way if a user want's a niche file format for an output, they can create it themselves.

**Reply by marrus-sh** [reply]
> XOXO is a 20 year old format that never caught on and isn't supported by any other outliner that I know of.

I agree with this, but I think “supported by other outliners” is a pretty narrow vision of cross‐compatibility. XOXO is, for example, supported by every browser, and every generic XML processing tool, whereas OMPL+Markdown is not supported by either. If Zavala works for me as an outliner, I don’t as a user care much about whether its export formats can work with other outliners, because I am already using Zavala. But I do care about whether its export formats can work well with other tools which *aren’t* outliners.

I don’t have any especial love for XOXO beyond it being a thing which has already been standardized which meets those requirements. I would be equally fine with extending OPML to provide the raw HTML data in the export in addition to the Markdown, for example. But because so many other things use OPML, those kinds of extensions seem more fraught. In this case, not having to worry about broad cross‐compatibility might be a benefit(?).

> You can work around the Markdown issue you mention using Shortcuts. Just convert the Markdown output from Zavala to Rich Text and then convert that Rich Text to HTML.

I will try this. A generic “rich text to HTML” conversion is going to be harder to work with than something more structured which more closely matches Zavala’s actual data model, but it might still be viable. I understand the desire to want to keep the application simple.

**Comment by vincode-io** [comment]
Would Microsoft's `.docx` format work for you?

**Reply by jsismail** [reply]
Notes compatibility is what I would like to see.  I have been using Notes to create outlines using indent/outdent (ie. not tabs). Would love to be able to paste a Zavala outline into Notes where indent/outdent formatting is used, and vice-versa — paste a note with indent/outdent formatting directly into Zavala.
